istream support for cfloat

Hello,

When reading numbers from a file to the fp16 format using `std::ifstream`, I'm getting an error:
~~~
undefined reference to `std::istream& sw::universal::operator>><16u, 5u, unsigned short, true, false, false>(std::istream&, sw::universal::cfloat<16u, 5u, unsigned short, true, false, false>&)'
~~~

I think this could be related to https://github.com/stillwater-sc/universal/blob/4c57ec1c9f8e07c064adb2cbb1433cea16c2f253/include/universal/number/cfloat/cfloat_impl.hpp#L3151

Any insights on this? If this would be hard to implement, how could I hack the reading fp16 numbers from a file?

Thanks!

--- a/include/sw/universal/number/cfloat/cfloat_fwd.hpp
+++ b/include/sw/universal/number/cfloat/cfloat_fwd.hpp
@@ -21,6 +21,10 @@ template<unsigned nbits, unsigned es, typename bt, bool hasSubnormals, bool hasM
 		cfloat<nbits, es, bt, hasSubnormals, hasMaxExpValues, isSaturating>
 		fabs(cfloat<nbits, es, bt, hasSubnormals, hasMaxExpValues, isSaturating>);
 
+// parsing
+template<unsigned nbits, unsigned es, typename bt, bool hasSubnormals, bool hasMaxExpValues, bool isSaturating>
+		bool parse(const std::string& number, cfloat<nbits, es, bt, hasSubnormals, hasMaxExpValues, isSaturating>& v);
+
 #ifdef CFLOAT_QUIRE
 
 // quire types

--- a/include/sw/universal/number/cfloat/cfloat_impl.hpp
+++ b/include/sw/universal/number/cfloat/cfloat_impl.hpp
@@ -22,6 +22,7 @@
 // 
 // supporting types and functions
 #include <limits>
+#include <regex>
 #include <type_traits>
 #include <universal/native/ieee754.hpp>
 #include <universal/native/subnormal.hpp>
@@ -3309,12 +3310,72 @@ inline std::ostream& operator<<(std::ostream& ostr, const cfloat<nbits, es, bt, 
 	return ostr << representation;
 }
 
-// istream input: currently marshalling through native double
+// parse a cfloat from a string in either cfloat hex format (nbits.esxHEXVALUEc)
+// or a decimal floating-point representation
+template<unsigned nbits, unsigned es, typename bt, bool hasSubnormals, bool hasMaxExpValues, bool isSaturating>
+bool parse(const std::string& txt, cfloat<nbits,es,bt,hasSubnormals,hasMaxExpValues,isSaturating>& v) {
+	// check if the txt is of the native cfloat form: nbits.esX[0x]hexvaluec
+	std::regex cfloat_regex(R"(^[0-9]+\.[0-9]+[xX](0[xX])?[0-9A-Fa-f]+c?$)");
+	if (std::regex_match(txt, cfloat_regex)) {
+		// found a cfloat representation: parse nbits.esxHEXVALUEc
+		std::string nbitsStr, esStr, bitStr;
+		auto it = txt.begin();
+		for (; it != txt.end(); ++it) {
+			if (*it == '.') break;
+			nbitsStr.append(1, *it);
+		}
+		for (++it; it != txt.end(); ++it) {
+			if (*it == 'x' || *it == 'X') break;
+			esStr.append(1, *it);
+		}
+		for (++it; it != txt.end(); ++it) {
+			if (*it == 'c') break;
+			bitStr.append(1, *it);
+		}
+		unsigned nbits_in = 0;
+		unsigned es_in = 0;
+		{
+			std::istringstream ss(nbitsStr);
+			ss >> nbits_in;
+			if (ss.fail()) return false;
+		}
+		{
+			std::istringstream ss(esStr);
+			ss >> es_in;
+			if (ss.fail()) return false;
+		}
+		// native cfloat form must match target configuration
+		if (nbits_in != nbits || es_in != es) return false;
+		uint64_t raw = 0;
+		std::istringstream ss(bitStr);
+		ss >> std::hex >> raw;
+		if (ss.fail()) return false;
+		ss >> std::ws;
+		if (!ss.eof()) return false;
+		v.setbits(raw);
+		return true;
+	}
+	else {
+		// assume it is a float/double/long double representation
+		std::istringstream ss(txt);
+		double d;
+		ss >> d;
+		if (ss.fail()) return false;
+		ss >> std::ws;
+		if (!ss.eof()) return false;
+		v = d;
+		return true;
+	}
+}
+
+// read an ASCII float or cfloat format: nbits.esxNN...NNc, for example: 16.5x7C00c
 template<unsigned nbits, unsigned es, typename bt, bool hasSubnormals, bool hasMaxExpValues, bool isSaturating>
 inline std::istream& operator>>(std::istream& istr, cfloat<nbits,es,bt,hasSubnormals,hasMaxExpValues,isSaturating>& v) {
-	double d(0.0);
-	istr >> d;
-	v = d;
+	std::string txt;
+	istr >> txt;
+	if (!parse(txt, v)) {
+		std::cerr << "unable to parse -" << txt << "- into a cfloat value\n";
+	}
 	return istr;
 }
 

--- a/static/float/cfloat/api/api.cpp
+++ b/static/float/cfloat/api/api.cpp
@@ -376,6 +376,49 @@ try {
 		std::cout << "polynomial(1.0) = " << polyeval(polynomial, 5, single(1.0f)) << '\n';
 	}
 
+	std::cout << "+------------ parsing ----------+\n";
+	{
+		int start = nrOfFailedTestCases;
+
+		// parse a decimal floating-point string
+		half a;
+		if (!parse("1.5", a)) ++nrOfFailedTestCases;
+		if (a != half(1.5)) ++nrOfFailedTestCases;
+
+		// parse a negative value
+		if (!parse("-3.25", a)) ++nrOfFailedTestCases;
+		if (a != half(-3.25)) ++nrOfFailedTestCases;
+
+		// parse scientific notation
+		if (!parse("1.25e3", a)) ++nrOfFailedTestCases;
+		if (a != half(1250.0)) ++nrOfFailedTestCases;
+
+		// parse a cfloat hex format: nbits.esxHEXVALUEc (with or without 0x prefix)
+		half b;
+		if (!parse("16.5x3C00c", b)) ++nrOfFailedTestCases;
+		if (b != half(1.0)) ++nrOfFailedTestCases;
+		// also accept hex_print format which includes 0x prefix
+		if (!parse("16.5x0x3C00c", b)) ++nrOfFailedTestCases;
+		if (b != half(1.0)) ++nrOfFailedTestCases;
+
+		// parse via istream operator>>
+		std::istringstream is("2.5");
+		half c;
+		is >> c;
+		if (c != half(2.5)) ++nrOfFailedTestCases;
+
+		// reject trailing junk
+		half d;
+		if (parse("1.5abc", d)) ++nrOfFailedTestCases;
+
+		// reject mismatched configuration
+		if (parse("32.8x12345678c", a)) ++nrOfFailedTestCases;
+
+		if (nrOfFailedTestCases - start > 0) {
+			std::cout << "FAIL : parsing\n";
+		}
+	}
+
 	ReportTestSuiteResults(test_suite, nrOfFailedTestCases);
 	return (nrOfFailedTestCases > 0 ? EXIT_FAILURE : EXIT_SUCCESS);
 }

